File: Algorithm Tempates/Prime Sieve.py
from math import sqrt


def get_primes_upto(n):
    """
        Uses Sieve of Eratosthenes to find 
        primes upto n 
    """
    isPrime = [1 for _ in range(n+5)]
    primes = []

    if n > 2:
        primes.append(2)

    limit = sqrt(n+0.0) + 2
    for i in range(3, n + 1, 2):
        if isPrime[i] == 1:
            primes.append(i)
            if i < limit:
                for j in range(i*i, n + 1, 2*i):
                    isPrime[j] = 0

    return primes


print(len(get_primes_upto(1000)))

File: Algorithm Tempates/Prime Sieve.cpp
#include <bits/stdc++.h>
#define SIZE 1000000

using namespace std;

bool isPrime[SIZE/2];
//only the mark of odd numbers is kept so total size is SIZE/2
vector <unsigned int> nPrime;
void primeSieve(int n)
{
    nPrime.push_back(2);

    int limit = sqrt(n+0.0) + 2;
    for(int i = 3;i <= n;i += 2){
        if( !isPrime[(i+1)/2] ){
            nPrime.push_back(i);
            if(i < limit){
                for(int j = i*i;j <= n;j += (i*2) ){
                    isPrime[(j+1)/2] = 1;
                    /*all composites below sqrt(i) is canceled
                     *so started canceling from i*i
                     *j += (2*i) because j+=i would give an even.
                     */
                }
            }
        }
    }
}

/*NB: Bit Sieve is more memory efficient.
 *because sizeof(bool) is not 1 bit,but 1 byte
 *because computer can't point memory < 1 byte
 */

int main(){
    primeSieve(100);
    for(int i = 0;i < nPrime.size();i++)
        cout << nPrime[i] << " ";
    cout << endl << "#primes: " << nPrime.size();
    return 0;
}

